### DISCUSSION TITLE
Send IDoc to SAP : Reference clientIDoc.py

### DISCUSSION THREAD

**Discussion by GreetyVarghese** [SAP/PyRFC#340] [Q&A] [answered]
Hi @bsrdjan , All, 
We have a requirement to send IDoc to SAP. Sample xml (1) is given below. Referring "https://github.com/SAP/PyRFC/blob/main/examples/clientIDoc.py", we have implemented python code in AWS Lambda (2). The Lambda implementation for 'idoc_data_dicts' is not complete as we are struggling to implement the nested structure here. For e.g., how to keep "E1OROPR " within "E1ORHDR " and also how to handle multiple levels of nested structures. 

Can you please help with sample implementations of IDocs. Thank You! 

1. IDoc to be sent to SAP 

```xml
<?xml version='1.0' encoding='UTF-8'?>
<Z_PM_IORDER01>
    <IDOC BEGIN="1">
        <EDI_DC40 SEGMENT="1">
            <TABNAM>EDI_DC40</TABNAM>
            <DIRECT>2</DIRECT>
            <IDOCTYP>IORDER01</IDOCTYP>
            <SNDPRT>LS</SNDPRT>
            <SNDPRN>MYDEVENV</SNDPRN>
            <RCVPOR>SAPA01</RCVPOR>
            <SNDPOR>MYDEVENV_S</SNDPOR>
            <RCVPRT>LS</RCVPRT>
            <RCVPRN>LOGA01C007</RCVPRN>
            <CREDAT>20231012</CREDAT>
            <CRETIM>20231012</CRETIM>
            <CIMTYP>Z_PM_IORDER01</CIMTYP>
            <MESTYP>IORDER</MESTYP>
            <MESFCT>NEW</MESFCT>
            <MESCOD>POC</MESCOD>
        </EDI_DC40>
        <E1ORHDR SEGMENT="1">
            <AUART>ABC1</AUART>
            <PRIOK>1</PRIOK>
            <ILART>B01</ILART>
            <E1OROPR SEGMENT="1">
                <LTXA1>Test Desc</LTXA1>
                <ZE1_PM_E1OROPR SEGMENT="1">
                    <USR00>SFDC_TEST</USR00>
                </ZE1_PM_E1OROPR>
            </E1OROPR>
            <E1OROPR SEGMENT="1">
                <LTXA1>Test Desc 1</LTXA1>
                <ZE1_PM_E1OROPR SEGMENT="1">
                    <USR00>SFDC_TEST</USR00>
                </ZE1_PM_E1OROPR>
            </E1OROPR>
            <E1OROPR SEGMENT="1">
                <LTXA1>Test Desc 2</LTXA1>
                <ZE1_PM_E1OROPR SEGMENT="1">
                    <USR00>SFDC_TEST</USR00>
                </ZE1_PM_E1OROPR>
            </E1OROPR>
            <E1OROPR SEGMENT="1">
                <LTXA1>Test Desc 3</LTXA1>
                <ZE1_PM_E1OROPR SEGMENT="1">
                    <USR00>SFDC_TEST</USR00>
                </ZE1_PM_E1OROPR>
            </E1OROPR>
            <ZE1ORHDR SEGMENT="1">
                <ZZ_DMP_ORD/>
                <RELKZ>X</RELKZ>
            </ZE1ORHDR>
            <E1ORTOB SEGMENT="1">
                <EQUNR>123456</EQUNR>
            </E1ORTOB>
        </E1ORHDR>
    </IDOC>
</Z_PM_IORDER01>`
```

2. Sample Lambda code 

```python
import json
import sys
from pyrfc import Connection
def get_idoc_desc(idoc_id):
    idoc_control = {
                'TABNAM': 'EDI_DC40',
                'DIRECT': '2',
                'IDOCTYP' : 'IORDER01',
                'SNDPRT' : 'LS',
                'SNDPRN' : 'MYDEVENV',
                'RCVPOR' : 'SAPA01',
                'SNDPOR' : 'MYDEVENV_S',
                'RCVPRT' : 'LS',
                'RCVPRN' : 'LOGA01C007',
                'CIMTYP' : 'Z_PM_IORDER01',
                'MESTYP' : 'IORDER',
                'MESFCT' : 'NEW',
                'MESCOD' : 'POC'
    }
    idoc_data_dicts={        
                 'AUART' : 'ABC1',
                 'PRIOK' : '1',
                 'ILART' : 'D01',
                #implement nested structure??
    }    
    new_idoc_data_dicts={
            "SEGNAM": "E1ORHDR",
            "SDATA" : json.dumps(idoc_data_dicts)
    }    
    return {
        "IDOC_CONTROL_REC_40": [idoc_control],
        "IDOC_DATA_REC_40": [new_idoc_data_dicts],
    }  
def lambda_handler(event, context):
    try:
        print("Connecting to SAP")  
        conn = Connection(ashost='a02.domain.com', sysnr='01', client='001', user='MYUSER', passwd='MYPSWD')
        print("Connected to SAP!!")
        # Define the IDOC
        idoc_id = 1
        idoc = get_idoc_desc(idoc_id)
        unit = conn.initialize_unit()
        print('Sending an IDOC to SAP')
        print(idoc)
        result=conn.fill_and_submit_unit( unit,[("IDOC_INBOUND_ASYNCHRONOUS", idoc)])
        print(f'IDOC Sent to SAP! and the response is : {str(result)}')
        conn.close()
    except Exception as e:
            print("Exception block!")
            print(f"An exception occurred: {str(e)}")
            return ("Failed")
    return ("Success!") `
```

**Comment by Ulrich-Schmidt** [comment, marked as answer]
Let me say this: trying to parse or create IDocs on the RFC level, is extremely difficult. At least when trying to do it in a generic way. The code for one particular IDoc (like your IORDER01) might be doable, but still is far from "simple"...

Just take a look at IDOC_INBOUND_ASYNCHRONOUS and its IDOC_DATA_REC_40 definition: it only has a 1000 CHAR field, into which the nested IDoc segment tree needs to be converted, and for this 

- you need to have "insider know-how" about the correct algorithm, how to fill the SEGNUM, PSGNUM and HLEVEL fields (I don't think it is publicly documented somewhere?!) and in which order these lines have to be added to the IDOC_DATA_REC_40 table, and
- you need to know the exact definition of each segment, so that you can calculate the offset inside the 1000 CHAR field at which the value for each segment-field needs to be inserted.

The last point is complicated by two factors:

1. The segment definition may change from SAP release to SAP release. If you develop and test your program against one system and then run it against a newer/older system (or your system gets upgraded), then the offsets may no longer be correct and you end up with "data salad". (Though this may more have been a problem 10-20 years ago. I think in the last 10 years, IDoc definitions for the standard main IDocs haven't been changed anymore.)
2. If the backend is non-Unicode, things get complicated by the fact that some CHARs may occupy 1 byte in the CHAR1000 field, and others may occupy 2 bytes. When the SDATA field is run through a codepage converter from non-Unicode to Unicode (as is the case for all three RFC connectors: the libsapnwrfc.so used by PyRFC, SAP JCo and SAP NCo), then the internal offsets and field-lengths are messed up quite unexpectedly, and again you end up with garbage data. In order to rectify this problem, you would have to interfere deep down on the protocol level (which you cannot do on the Python level) and split the 1000 **Byte** field according to the segment's offsets and field-lengths, **before** it is codepage-converted, and then convert the bytes to Unicode **Chars** on a field-by-field basis...

So if you want my advise: if you plan on processing more than 2-3 IDoc types, don't even try...

If you have the chance to do this in Java, then you can use [SAP's JCo IDoc Library](https://support.sap.com/en/product/connectors/jco.html?anchorId=section_754631833). It does all the hard work for you! (And I'm not exaggerating when I say that around two years of net development time went into this library...)

Another idea I just had, as it looks like you are receiving the IDoc data in SAP's standard IDoc-XML format: have you taken a look at [SAP Business Connector](https://support.sap.com/en/product/connectors/bc.html) yet? It can act as an HTTP(S) server, receive the IDoc-XML via HTTP POST, automatically convert it to an IDoc and send it into an SAP backend system. If this is your scenario, you could achieve it without having to code a single line. You would only need a bit of configuration like this:

- Install SAP BC
- Configure an HTTP listening port (possibly with credentials/certificates that your partner can use for authenticating)
- Define connection parameters for the connected SAP Backend System
- Define a Routing Rule (or a couple of them, if different IDoc types have to be routed to different SAP Backends)
- And you are done!

(In addition to that, you also get a "monitoring component" for free, where you can monitor failed IDocs, e.g. because they had problems during HTTP or RFC communication. The opposite direction, SAP (IDoc) --> SAP BC -->  HTTP receiver(IDoc-XML) is also supported out-of-the-box.)
